Adding autocasting (#576)
crystal-lang/crystal-book: docs/syntax_and_semantics/autocasting.md

# Type autocasting

> At function calls, Crystal transparently casts elements of certain types when there is no ambiguity.

Reviewer: Oh, I was wrong there. Autocasting is not limited to function calls. It also applies to assignments. Maybe I missed something else?
```suggestion
Crystal transparently casts elements of certain types when there is no ambiguity.
```

Author: I added an example for the other situation where this works.

## Number autocasting

> When passing a number to a function, values of a type are autocasted to a larger one if no precision is lost:

Reviewer: ```suggestion
Values of a numeric type autocast to a larger one if no precision is lost:
```

```crystal
def foo(x : Int32) : Int32
  x
end

def bar(x : Float32) : Float32
  x
end

def bar64(x : Float64) : Float64
  x
end

foo 0xFFFF_u16 # OK, an UInt16 always fit an Int32
foo 0xFFFF_u64 # OK, this particular Int64 fit in an Int32
bar(foo 1)     # Fails, casting an Int32 to a Float32 might lose precision
bar64(bar 1)   # OK, a Float32 can be autocasted to a Float64
```

> Number literals are always casted when the actual value of the literal fits the target type, despite of its type. Expressions are casted too (like in the last example above), unless the flag `no_number_autocast` is passed to the compiler (see [Compiler features](compile_time_flags.md#compiler-features)).

Reviewer: I think the last sentence should be moved to a separate paragraph, it's unrelated to autocasting of literals.
